firmata/index.d.ts

```diff
@@ -0,0 +1,303 @@
+// Type definitions for firmata.js v0.15.0
+// Project: https://github.com/firmata/firmata.js
+// Definitions by: Troy W. <https://github.com/troywweber7, https://bitbucket.org/troywweber/>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+declare module 'firmata'
+{
+	import * as SerialPort from 'serialport'
+	export = Board;
+
+	/**
+	 * Most of these are generated by observing https://github.com/firmata/firmata.js/blob/master/lib/firmata.js.
+	 *
+	 * This is a starting point that appeared to work fine for months within a project of my company, but I give no
+	 * guarantee that it cannot be improved.
+	 */
+	class Board extends NodeJS.EventEmitter
+	{
+		constructor(serialPort:string, callback?:(error:any) => void)
+		public MODES:Board.IPinModes;
+		public STEPPER:Board.IStepperConstants;
+		public I2C_MODES:Board.II2cModes;
+		public SERIAL_MODES:Board.ISerialModes;
+		public SERIAL_PORT_IDs:Board.ISerialPortIds;
+		public SERIAL_PIN_TYPES:Board.ISerialPinTypes;
+		public HIGH:Board.PIN_STATE;
+		public LOW:Board.PIN_STATE;
+		public pins:Board.IPins[];
+		public ports:number[];
+		public analogPins:number[];
+		public version:Board.IVersion;
+		public firmware:Board.IFirmware;
+		public settings:Board.ISettings;
+		public reportVersion(callback:() => void):void
+		public queryFirmware(callback:() => void):void
+		public analogRead(pin:number, callback:(value:number) => void):void
+		public analogWrite(pin:number, value:number):void
+		public pwmWrite(pin:number, value:number):void
+		public servoConfig(pin:number, min:number, max:number):void
+		public servoWrite(pin:number, value:number):void
+		public pinMode(pin:number, mode:Board.PIN_MODES):void
+		public digitalWrite(pin:number, val:Board.PIN_STATE):void
+		public digitalRead(pin:number, callback:(val:Board.PIN_STATE) => void):void
+		public queryCapabilities(callback:() => void):void
+		public queryAnalogMapping(callback:() => void):void
+		public queryPinState(pin:number, callback:() => void):void
+		// TODO untested/incomplete --- TWW
+		public sendString(str:string):void
+		// TODO untested/incomplete --- TWW
+		public sendI2CConfig(delay:number):void
+		// TODO untested/incomplete --- TWW
+		public i2cConfig(options:any):void
+		// TODO untested/incomplete --- TWW
+		public sendI2CWriteRequest(slaveAddress:any, bytes:any):void
+		// TODO untested/incomplete --- TWW
+		public i2cWrite(address:any, registerOrData:any, inBytes:any):void
```

Comment:
RESOLVED no idea if that will provide auto-completion for either/or, I think it will